File: Universities ICPC/iut/I2.cpp
#include <bits/stdc++.h>

using namespace std;
int M,N;
int mat[507][507];
int aux[507][507];
void preProcess()
{
   for (int i=0; i<N; i++)
      aux[0][i] = mat[0][i];
 
   for (int i=1; i<M; i++)
      for (int j=0; j<N; j++)
         aux[i][j] = mat[i][j] + aux[i-1][j];
 
   for (int i=0; i<M; i++)
      for (int j=1; j<N; j++)
         aux[i][j] += aux[i][j-1];
}
int sumQuery(int tli, int tlj, int rbi,int rbj)
{

    int res = aux[rbi][rbj];
     if (tli > 0)
       res = res - aux[tli-1][rbj];
 
    if (tlj > 0)
       res = res - aux[rbi][tlj-1];
 
    if (tli > 0 && tlj > 0)
       res = res + aux[tli-1][tlj-1];

    return res;
}
int main()
{
    cin>>M>>N;
    for (int i = 0; i < M; i++)
    {
        for (int j = 0; j < N; j++)
        {
            cin>>mat[i][j];
        }
    }
    int T; cin>>T;
    int tli,tlj,rbi,rbj;
    while (T--)
    {
        cin>>tli>>rbi>>tlj>>rbj;
        cout<<sumQuery(tli,tlj,rbi,rbj)<<'\n';
    }
}

File: Universities ICPC/iut/I2.py
M,N = map(int,input().split())

def preProcess(mat, aux):
     
    for i in range(0, N, 1):
        aux[0][i] = mat[0][i]
 
    for i in range(1, M, 1):
        for j in range(0, N, 1):
            aux[i][j] = mat[i][j] + aux[i - 1][j]
 
    for i in range(0, M, 1):
        for j in range(1, N, 1):
            aux[i][j] += aux[i][j - 1]
 
def sumQuery(aux, tli, tlj, rbi, rbj):
    res = aux[rbi][rbj]
    if (tli > 0):
        res = res - aux[tli - 1][rbj]
 
    if (tlj > 0):
        res = res - aux[rbi][tlj - 1]
        
    if (tli > 0 and tlj > 0):
        res = res + aux[tli - 1][tlj - 1]
 
    return res

mat = []
for i in range(M):
    temp = list(map(int,input().split()))
    mat.append(temp)

aux = [[0 for i in range(N)] for j in range(M)]
preProcess(mat, aux)

T = int(input())

for i in range(T):
    tli,rbi,tlj,rbj =map(int,input().split())
    print(sumQuery(aux, tli, tlj, rbi, rbj))

